### PR TITLE
refactor and tests in order to allow returning dependent fields for query

### PR DESCRIPTION
**Pull request by nicpottier**
Bit of shuffling about to make it so we can get the fields that make up a query and pass those back. (so rapid knows whether dynamic groups are possible).  Number of return values was getting a bit bananas so pass around the ContactQL query around now.

Need to address https://github.com/nyaruka/goflow/issues/834 before merging.

### CODE REVIEW

File: search/search.go [hunk 1]
```diff
@@ -13,19 +14,54 @@ import (
...
 	if err != nil {
-		return "", nil, NewError(err.Error())
+		return nil, NewError(err.Error())
+	}
+
+	return eq, nil
+}
+
+// DependentFields returns all the dependent fields for the passed in query. This includes attributes such as "id" and "name"
+func DependentFields(query *contactql.ContactQuery) []string {
```
Comment (rowanseymour) [inline search/search.go]: "Dependent" is sorta wrong way around

File: web/contact/contact.go [hunk 2]
```diff
@@ -82,8 +83,9 @@ func handleSearch(ctx context.Context, s *web.Server, r *http.Request) (interfac
 
 	// build our response
 	response := &searchResponse{
-		Query:      parsed,
+		Query:      parsed.String(),
```
Comment (rowanseymour) [inline web/contact/contact.go:86]: assume this is waiting on https://github.com/nyaruka/goflow/issues/834

File: models/contacts.go [hunk 3]
```diff
@@ -175,26 +175,32 @@ func ParseQuery(org *OrgAssets, resolver contactql.FieldResolverFunc, query stri
...
-		return "", nil, 0, errors.Errorf("no elastic client available, check your configuration")
+		return nil, nil, 0, errors.Errorf("no elastic client available, check your configuration")
 	}
 
 	resolver := buildFieldResolver(org)
-	parsed, eq, err := ParseQuery(org, resolver, query)
+
+	parsed, err := search.ParseQuery(org.Env(), resolver, query)
+	if err != nil {
+		return nil, nil, 0, errors.Wrapf(err, "error parsind query: %s", query)
```
Comment (rowanseymour) [inline models/contacts.go]: typo

File: search/search_test.go [hunk 4]
```diff
@@ -76,24 +84,40 @@ func TestElasticSort(t *testing.T) {
...
+func TestQueryTerms(t *testing.T) {
+	resolver := buildResolver()
+
+	tcs := []struct {
+		Query  string
+		Fields []string
+	}{
+		{"joe", []string{"name"}},
+		{"id = 10", []string{"id"}},
+		{"name = joe or age > 10", []string{"age", "name"}},
```
Comment (rowanseymour) [inline search/search_test.go]: can't remember if everything is lowercased at this point but worth adding a test where it isn't

**Review by rowanseymour — APPROVED** [review]

**Comment by nicpottier** [comment]
Updated @rowanseymour .. so this includes the goflow webhook changes you made as well, do those need to be coordinated with anything on the rapidpro side or can I merge this?

**Comment by rowanseymour** [comment]
@nicpottier I think it should behave exactly the same unless you change MaxBodyBytes on the engine.. but I think we should merge https://github.com/nyaruka/goflow/pull/835 before doing that.

**Pull request merged**